### PR TITLE
Opensource examples

### PR DESCRIPTION
**Pull request by stoeffel**
Co-authored-by: jwoudenberg-nri <jasper@noredink.com>
Co-authored-by: michaelglass <mike@noredink.com>

related PR https://github.com/NoRedInk/NoRedInk/pull/33471/files

### CODE REVIEW

**Comment by stoeffel** [comment]
todo
- [ ] opensource Nri.WebServer.TestEncoding as well.

File: nri-prelude/src/Examples.hs [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+-- APIs. This allows us to write tests that will warn us when the encoding of
+-- our types change, potentially in backwards-incompatible ways.
+module Examples
+  ( HasExamples (..),
+    Example (..),
+    example,
+    concat,
+    toList,
+    toFileName,
+    toTestName,
```
Comment (jwoudenberg-nri) [inline nri-prelude/src/Examples.hs:14]: :thought_balloon: `toFileName` and `toTestName` feel to me like parts of the implementation of `Example` tests. Shall we drop them from here?

File: nri-prelude/src/Examples.hs [hunk 2]
```diff
@@ -0,0 +1,101 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE PolyKinds #-}
+
+-- | Helpers for associating example values with all the types we use in our
+-- APIs. This allows us to write tests that will warn us when the encoding of
+-- our types change, potentially in backwards-incompatible ways.
+module Examples
```
Comment (jwoudenberg-nri) [inline nri-prelude/src/Examples.hs:7]: :heart_eyes_cat:

File: nri-prelude/src/Examples.hs [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+
+-- | A helper type class that provides us example values of particular types.
+-- The `IsApi` typeclass below will demand we define an instance of this type
+-- class for each type used in a request or response body.
+class HasExamples t where
+  examples :: Proxy t -> (Example, List Example)
+
+-- | Concat two "nonempty" lists of examples together.
+concat :: (a, List a) -> (a, List a) -> (a, List a)
+concat (x, xs) (y, ys) = (x, xs ++ (y : ys))
```
Comment (jwoudenberg-nri) [inline nri-prelude/src/Examples.hs:53]: I think I have an idea that would make it so we don't need to expose these internals. Going to give it a try on a branch and ask for your feedback.

File: nri-test-encoding/README.md [hunk 4]
```diff
@@ -0,0 +1,4 @@
+# NriTestEncoding
```
Comment (jwoudenberg-nri) [inline nri-test-encoding/README.md:1]: :tada:

**Review by jwoudenberg-nri — APPROVED** [review]
Looks great!

**Pull request merged**